router/core/context.go

```diff
@@ -573,6 +578,47 @@ func (o *operationContext) ClientInfo() ClientInfo {
 	return *o.clientInfo
 }
 
+type QueryPlanStats struct {
+	TotalSubgraphFetches int
+	SubgraphFetches      map[string]int
+}
+
+func (p *QueryPlanStats) analyzePlanNode(plan *resolve.FetchTreeQueryPlanNode) {
+	switch plan.Kind {
+	case resolve.FetchTreeNodeKindSingle:
+		p.analyzeSingleFetch(plan)
+	case resolve.FetchTreeNodeKindSequence, resolve.FetchTreeNodeKindParallel:
+		for _, child := range plan.Children {
+			p.analyzePlanNode(child)
+		}
+	}
+}
+
+func (p *QueryPlanStats) analyzeSingleFetch(plan *resolve.FetchTreeQueryPlanNode) {
+	key := plan.Fetch.SubgraphName
+
+	p.TotalSubgraphFetches++
+
+	if entry, ok := p.SubgraphFetches[key]; ok {
+		p.SubgraphFetches[key] = entry + 1
+	} else {
+		p.SubgraphFetches[key] = 1
+	}
+}
+
+func (o *operationContext) QueryPlanStats() QueryPlanStats {
+	qps := QueryPlanStats{
+		TotalSubgraphFetches: 0,
+		SubgraphFetches:      make(map[string]int),
+	}
+
+	if p, ok := o.preparedPlan.preparedPlan.(*plan.SynchronousResponsePlan); ok {
+		qps.analyzePlanNode(p.Response.Fetches.QueryPlan())
```

Comment:
Can QueryPlan ever return nil?

Reply:
I'd suggest to do a nil check to be on the safe side. People can call this currently from several places.

Reply:
If they do call it from a place where it completely fails, is it better to fail silently with an empty QPS? Maybe I should make it return (QPS, error)?

Reply:
Exactly, cool thx